naga/src/back/glsl/mod.rs

```diff
@@ -2620,51 +2610,49 @@ impl<'a, W: Write> Writer<'a, W> {
                 level,
                 depth_ref,
             } => {
-                let dim = match *ctx.resolve_type(image, &self.module.types) {
-                    TypeInner::Image { dim, .. } => dim,
+                let (dim, class, arrayed) = match *ctx.resolve_type(image, &self.module.types) {
+                    TypeInner::Image {
+                        dim,
+                        class,
+                        arrayed,
+                        ..
+                    } => (dim, class, arrayed),
                     _ => unreachable!(),
                 };
-
-                if dim == crate::ImageDimension::Cube
-                    && array_index.is_some()
-                    && depth_ref.is_some()
-                {
-                    match level {
-                        crate::SampleLevel::Zero
-                        | crate::SampleLevel::Exact(_)
-                        | crate::SampleLevel::Gradient { .. }
-                        | crate::SampleLevel::Bias(_) => {
-                            return Err(Error::Custom(String::from(
-                                "gsamplerCubeArrayShadow isn't supported in textureGrad, \
-                                 textureLod or texture with bias",
-                            )))
-                        }
-                        crate::SampleLevel::Auto => {}
+                let mut err = None;
+                if dim == crate::ImageDimension::Cube {
+                    if offset.is_some() {
+                        err = Some("gsamplerCube[Array][Shadow] doesn't support texture sampling with offsets");
+                    }
+                    if matches!(level, crate::SampleLevel::Gradient { .. }) {
+                        err = Some("samplerCubeArrayShadow don't support textureGrad");
                     }
                 }
+                if gather.is_some() && level != crate::SampleLevel::Zero {
+                    err = Some("textureGather doesn't support LOD parameters");
+                }
+                if dim == crate::ImageDimension::D3 && arrayed {
+                    err = Some("3d array textures don't support sampling");
+                }
```

Comment:
The validator usually implements the WGSL restrictions and use [`Capabilities`](https://github.com/gfx-rs/wgpu/blob/07e59eb6fc7de3f682f1c401b9cf9f0da9ee4b4a/naga/src/valid/mod.rs#L80) to allow extra things through.

In this case, I'm not sure how prevalent usage of 1d storage arrays is. I think wgpu also won't allow you to create 1d-array texture views. It might be best to validate out 1d arrays (https://github.com/gfx-rs/wgpu/issues/4405) and if people would like to use them in the future we implement proper support for them.

Since https://github.com/gfx-rs/wgpu/issues/4405 is already tracking this, it can be a different PR.